### PR TITLE
Allow overriding features via env

### PR DESCRIPTION
**Pull request by alexcb**
Signed-off-by: Alex Couture-Beil <alex@earthly.dev>

### CODE REVIEW

File: features/features.go [hunk 1]
```diff
@@ -18,7 +18,8 @@ import (
 type Features struct {
 	// DoSaves flags the feature to only save artifacts under directly referenced targets
 	// of those referenced by BUILDs
-	ReferencedSaveOnly bool `long:"referenced-save-only" description:"only save artifacts that are directly referenced"`
+	ReferencedSaveOnly     bool `long:"referenced-save-only" description:"only save artifacts that are directly referenced"`
+	UseCopyIncludePatterns bool `long:"use-copy-include-patterns" description:"specify an include pattern to buildkit when performing copies"`
```
Comment (alexcb) [inline features/features.go:22]: This isn't currently referenced anywhere else; however, I needed a second flag for unit testing the comma splitting.

Once this PR is in, i'll rebase https://github.com/earthly/earthly/pull/1085 against this code which will then make use of this field for feature flipping.

File: earthfile2llb/earthfile2llb.go [hunk 2]
```diff
@@ -105,6 +106,11 @@ func Earthfile2LLB(ctx context.Context, target domain.Target, opt ConvertOpt) (m
 		return nil, errors.Wrapf(err, "resolve feature set for version %v for target %s", bc.Earthfile.Version.Args, target.String())
 	}
 
+	err = features.ApplyFlagOverrides(ftrs, os.Getenv("EARTHLY_VERSION_FLAGS"))
```
Comment (alexcb) [inline earthfile2llb/earthfile2llb.go:109]: unsure if we should call this `EARTHLY_VERSION_FLAGS` or `EARTHLY_VERSION_FLAG_OVERRIDES`

Comment (dchw) [reply earthfile2llb/earthfile2llb.go:109]: Would this make more sense next to the rest of our config in `main.go`?

**Review by dchw — APPROVED** [review]

**Comment by dchw** [comment]
What is our intended lifetime of these feature flags? Just hoping we dont accumulate too many that stick around for too long

File: earthfile2llb/earthfile2llb.go (hunk 2 above)
Comment (alexcb) [reply earthfile2llb/earthfile2llb.go:109]: That's a good point, I can leave it Hidden as it's only (mostly?) a dev feature.

**Comment by alexcb** [comment]
> What is our intended lifetime of these feature flags? Just hoping we dont accumulate too many that stick around for too long

Good question, under https://github.com/earthly/earthly/issues/991 we outlined a rollout strategy for the version numbering. I suspect the feature flags will be treated in a similar way:

> We aim to support at least 4 minor versions behind and also at least a whole year since the last release of that minor version. This should be a nice balance between keeping Earthly development nimble enough, while not breaking people's builds abruptly.

File: earthfile2llb/earthfile2llb.go (hunk 2 above)
Comment (dchw) [reply earthfile2llb/earthfile2llb.go:109]: 100% was assuming it would be hidden.

**Comment by dchw** [comment]
So the flags will stay supported for as long as the version they were introduced in stay supported. Feels fair. :+1:

File: cmd/earthly/main.go [hunk 3]
```diff
@@ -563,6 +564,13 @@ func newEarthlyApp(ctx context.Context, console conslogging.ConsoleLogger) *eart
 			Usage:       "Disable collection of analytics",
 			Destination: &app.disableAnalytics,
 		},
+		&cli.StringFlag{
+			Name:        "feature-flag-overrides",
+			EnvVars:     []string{"EARTHLY_FEATURE_FLAG_OVERRIDES"},
+			Usage:       "Apply additional flags after the VERSION command is parsed, multiple flags can be seperated by commas",
+			Destination: &app.featureFlagOverrides,
+			Hidden:      true, // used for feature-flipping from ./earthly dev script
+		},
```
Comment (vladaionescu) [inline cmd/earthly/main.go:573]: Would these apply to all Earthfiles encountered during the build?

**Review by vladaionescu — APPROVED** [review]
:+1:

File: cmd/earthly/main.go (hunk 3 above)
Comment (alexcb) [reply cmd/earthly/main.go:573]: yes, does something like `Override all VERSION commands with these additional flags, multiple flags can be seperated by commas` make it more clear?

Comment (vladaionescu) [reply cmd/earthly/main.go:573]: Yes - that makes sense

**Pull request merged**